Fix for service backend entry leak in cilium_lb4_backends_v2

hemanthmalla opened the pull request:
Backends are meant to be reused across services. When service backends are updated, existence of every backend is looked up by it's hash. Currently, the check is made only against the current service's backends but not against global backends.

This can result in duplicate backend entries. Without this commit whether duplicate is created or not depends on a race between releasing old backend ID during deletion and updating ref. count during creation. Currently on agent startup when service restoration process finds a duplicate entry, restoration stops and the rest of the entries are leaked. Over time with enough restarts, we could max out the backend bpf map.

This commit makes the lookup global and entries in backendByHash map are deleted only after data from lbmap is deleted. This should prevent duplicate entries in backend map.

Fixes #23551

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

aditighag reviewed (changes requested):
https://github.com/cilium/cilium/pull/23749#issuecomment-1442672185

hemanthmalla committed:
Defensive checks to avoid potential service backend entry leaks

Follow up for #23858 . During backend deletion, reference count
decrement and backend ID release operations could go through, but
deletion of entries from bpf map might fail. This commit adds
additional defenses to avoid potential leaks or duplicates from
similar scenarios in the future.

Also see https://github.com/cilium/cilium/pull/23749#issuecomment-1435082117 for more details on past race conditions.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

aditighag commented on pkg/service/service.go:
```diff
@@ -1555,21 +1555,28 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []*lb.Backend
-
 		if b, found := svc.backendByHash[hash]; !found {
-			if s.backendRefCount.Add(hash) {
+			globalBackend, existsGlobally := s.backendByHash[hash]
```
Umm... When there is a leak, datapath state goes out of sync with userspace. Will this work when a backend is associated with only one service?

hemanthmalla replied:
My understanding was that `s.backendByHash` would still have the entry because its deleted only after the entry is removed from the lbmap. Since the backend ID might be released I'm attempting to restore if an entry exists in `s.backendByHash`
https://github.com/cilium/cilium/blob/962934301d11ea58007ed6d69c7835d45899450f/pkg/service/service.go#L1391-L1399

aditighag replied:
I completely agree with the motivation to prevent such bugs from happening in future. The fix looks fine based on the current logic, but we are relying on the ordering of release of backend id and deletion from the lbmap. Could you add unit tests?

aditighag reviewed (commented):
Sorry for the delay! Thanks for the revised fix.
